### DISCUSSION TITLE
Correct approach of decoding custom binary formats using providers?

### DISCUSSION THREAD

**Discussion by gotthardp** [openssl/openssl#22627] [Q&A]
Hello. I want to load (decode) a custom binary format that stores a TPM2 context, so it can be used as a private key. The file format is not DER, MSBLOB nor PVK. It's a binary structure defined by the tpm2-tools project. I implemented the appropriate decoders and put them to the `OSSL_OP_DECODER` list as follows:
```
static const OSSL_ALGORITHM tpm2_decoders[] = {
    { "TSSCTX", "provider=tpm2,input=tssctx", tpm2_ctx_decoder_functions },
    { "RSA", "provider=tpm2,input=tssctx,structure=TPMS_CONTEXT", tpm2_ctx_to_rsa_decoder_functions },
    { "EC", "provider=tpm2,input=tssctx,structure=TPMS_CONTEXT", tpm2_ctx_to_ec_decoder_functions },
    { NULL, NULL, NULL }
};
```
But the `TSSCTX` decoder is never called. It is likely because in `file_store.c` the `file_setup_decoders()` first adds `ossl_any_to_obj_algorithm` for DER, MSBLOB and PVK and then calls `OSSL_DECODER_CTX_add_extra` that *goes through existing decoder methods what the existing ones want as input, and push those newly fetched decoders on top of the same stack.* Since at this moment only DER, MSBLOB or PVK is wanted as an input the `TSSCTX` decoder is never fetched.

I made everything to work by "abusing" one of the existing formats. Everything works as expected when I pretend to be using e.g. the MSBLOB and do:
```
static const OSSL_ALGORITHM tpm2_decoders[] = {
    { "MSBLOB", "provider=tpm2,input=msblob", tpm2_ctx_decoder_functions },
    { "RSA", "provider=tpm2,input=msblob,structure=TPMS_CONTEXT", tpm2_ctx_to_rsa_decoder_functions },
    { "EC", "provider=tpm2,input=msblob,structure=TPMS_CONTEXT", tpm2_ctx_to_ec_decoder_functions },
    { NULL, NULL, NULL }
};
```

Although it works it feels wrong, because it's not really the MSBLOB. Is/should there be a better way of defining custom binary decoders, please?

**Comment by t8m** [comment]
This sounds like a bug. But @levitte might want to chime in.
